rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,416 @@
+unless there are new requests from framework users.
+
+### Multiple devices
+For multiple devices scenarios, users can still use the `void
+set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, size_t
+capacity)` API to set capacity. But the specified capacity here is the
+budget for each device instead of a total budget for all devices. And the
+library will maintain dedicated cache for each device and manage them
+separately.
+
+The above proposal is based on a consideration: only limiting total budget for
+all devices may lead to a situation that one device has been OOM since cached
+too much constant tensor but the whole cache size still does not reach the
+capacity.
+
+If users want to set different capacity for different device, we need to let
+users specify which device to be set through API.
+
+oneDNN Graph API allows users to use multiple devices in two ways:
+- create engine with a device `index`.
+- create engine with a `sycl::device`.
+
+So, the following frontend capacity control APIs can be added to specify constant
+tensor cache capacity for each device:
+
+```cpp
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int index, size_t size);
+void get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int index, size_t* size);
+
+// The sycl interop API
+void sycl_interop::set_constant_tensor_cache_capacity(const sycl::device & dev, size_t size);
+void sycl_interop::get_constant_tensor_cache_capacity(const sycl::device & dev, size_t* size);
+```
+
+The weight sharing rules in multiple devices scenarios
+
+In some special cases, like dynamic shape, multiple different compiled
+partitions may share same weight. Based on the different combination of device
+and context used by each compiled partition, we have some general rules:
+- same device, same context: the compiled partitions that use both same device
+  and context are allowed to share same cached weight.
+- same device, different context: sycl programming api doesn't allow access
+  memory cross different context. So the compiled partitions that use different
+  context are not allowed to share same cached weight, even them are using same
+  device.
+- different device, same context: based on above discussions, constant tensor
+  cache is device specific, so the compiled partitions that use different device
+  will lookup different cache and will not hit same cached weight. This makes
+  sense, because accessing memory cross device is usually slower than accessing
+  local memory, so it's not preferred.
+- different device, different context: not sharing weight at all.
+
+Based on above analysis, weight sharing cross compiled partitions will only
+happens when those compiled partitions are using same device and same context.
+
+Supporting multiple devices usage remains an open topics. Any suggestions or
+ideas are welcome and we can discuss it in more details once real requests
+emerged.
```

Comment:
We may need to temporary conclusion here to proceed the implementation. As so far we do not see the requirement or a real case for setting different cache sizes for multiple different cards, I would suggest to keep the existing design by setting the size control per engine kind. We can extend the API in future once the requirement pops up.

Reply:
updated to: Currently, to make the API simple, we prefer to still use the `void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, size_t capacity)` API to set same capacity for all devices of same engine kind. We can extend the API in future once the requirement pops up.